### PR TITLE
Enhance Transaction Simulation: Bypass Auth Key Check and Fee Payment, Improve Multisig Consistency

### PR DESCRIPTION
**Pull request by junkil-park**
## Description

This PR enhances the simulation functionality in several key ways:

1. Bypass Authentication Key Check:
   * Removes the mandatory check for an authentication key during simulations.
   * This addresses a long-standing and frequently requested feature in [this issue](https://github.com/aptos-labs/aptos-core/issues/6862).

2. Eliminate Gas Fee Payment Requirement:
   * Removes the requirement for a gas fee payer during simulations when the fee payer address is 0x0.
   * Simulations will no longer validate the presence of sufficient funds for gas payment, allowing developers to test scenarios without considering gas fees.
   * This addresses the feature request in [this issue](https://github.com/aptos-labs/aptos-core/issues/13686)

3. Improve Multisig Simulation Consistency:
   * Merges the Multisig payload simulation path with the execution path, ensuring consistency between simulation and actual execution.
   * This resolves the issue of an onchain payload not being retrived (desribed [here](https://github.com/aptos-labs/aptos-core/issues/12703) and [here](https://github.com/aptos-labs/aptos-core/issues/8304)).
   * This also resolves the issue of inaccurate gas estimation for multisig transactions (described [here](https://github.com/aptos-labs/aptos-core/issues/12704)).

These changes are aimed at improving the flexibility of the simulation environment, allowing developers to test transactions and interactions without the constraints of authentication keys and gas fee payments.

This PR resolves, https://github.com/aptos-labs/aptos-core/issues/6862, https://github.com/aptos-labs/aptos-core/issues/13686, https://github.com/aptos-labs/aptos-core/issues/12703, https://github.com/aptos-labs/aptos-core/issues/8304, https://github.com/aptos-labs/aptos-core/issues/12704.

## Type of Change
- [X] New feature
- [X] Bug fix
- [X] Breaking change
- [ ] Performance improvement
- [X] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [x] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [X] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [x] Aptos Framework
- [X] Aptos CLI/SDK
- [X] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
e2e API tests and tested on the localnet

### CODE REVIEW

File: scripts/authenticator_regenerate.sh [hunk 1]
```diff
@@ -37,19 +37,6 @@ cd $repodir
...
-  echo "Regenerating Typescript SDK (in `pwd`)"
-  # Typescript SDK client files
-  cd ecosystem/typescript/sdk
-  pnpm install
-  pnpm generate-client
-
-  # Typescript SDK docs
-  pnpm generate-ts-docs
-  cd ..
-)
```
Comment (gregnazario) [inline scripts/authenticator_regenerate.sh:51]: Why is this removed btw?  Though, I guess the SDK doesn't need to be built anyways, it'll be deprecated.

File: ecosystem/typescript/sdk/examples/typescript/multisig_account.ts [hunk 2]
```diff
@@ -83,12 +83,6 @@ const { AccountAddress, EntryFunction, MultiSig, MultiSigTransactionPayload, Tra
   const multisigTxExecution = new TransactionPayloadMultisig(
     new MultiSig(AccountAddress.fromHex(multisigAddress), transferTxPayload),
   );
-  // We can simulate the transaction to see if it will succeed without having to create it on chain.
-  const [simulationResp] = await client.simulateTransaction(
-    owner2,
-    await client.generateRawTransaction(owner2.address(), multisigTxExecution),
-  );
-  assert(simulationResp.success);
```
Comment (gregnazario) [inline ecosystem/typescript/sdk/examples/typescript/multisig_account.ts:91]: We don't want to keep this?

**Review by gregnazario — COMMENTED** [review]
Can we double check that you can't submit a NoSignature transaction and have it pass?  Do we have new tests that cover this new path?

File: scripts/authenticator_regenerate.sh (hunk 1 above)
Comment (junkil-park) [reply scripts/authenticator_regenerate.sh:51]: `pnpm generate-ts-docs` gives the following error, and Maayan also told me that we don't need to update it.
```
Cloning into '/tmp/ts-sdk-doc'...
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

File: ecosystem/typescript/sdk/examples/typescript/multisig_account.ts (hunk 2 above)
Comment (junkil-park) [reply ecosystem/typescript/sdk/examples/typescript/multisig_account.ts:91]: This is the old way to simulate a multisig transaction before creation. After this PR, one should simulate multisig simulations using multisig accounts as the sender without the auth key check. I will update the TS examples accordingly once the SDK supports `NoAccountSignature`.

File: api/src/tests/multisig_transactions_test.rs [hunk 3]
```diff
@@ -411,13 +411,17 @@ async fn test_multisig_transaction_simulation() {
         .create_multisig_account(
             owner_account_1,
             vec![owner_account_2.address(), owner_account_3.address()],
-            2,    /* 2-of-3 */
```
Comment (banool) [inline api/src/tests/multisig_transactions_test.rs:414]: Are we testing the 2 of 3 case anywhere now?

File: api/src/tests/simulation_test.rs [hunk 4]
```diff
@@ -153,3 +157,200 @@ async fn test_simulate_txn_with_aggregator() {
...
+    let txn = context.mint_user_account(alice).await;
+    context.commit_block(&vec![txn]).await;
+
+    // Construct a signed transaction.
+    let txn = context.account_transfer_to(alice, bob.address(), transfer_amount);
+    // Replace the authenticator with a NoAccountAuthenticator in the transaction.
+    let txn = SignedTransaction::new_signed_transaction(
+        txn.clone().into_raw_transaction(),
+        TransactionAuthenticator::SingleSender {
+            sender: AccountAuthenticator::NoAccountAuthenticator,
```
Comment (banool) [inline api/src/tests/simulation_test.rs:223]: We need a test that proves that this variant is rejected for the normal txn submission path.

File: aptos-move/aptos-release-builder/src/components/feature_flags.rs [hunk 5]
```diff
@@ -121,6 +121,7 @@ pub enum FeatureFlag {
     DefaultToConcurrentFungibleBalance,
     LimitVMTypeSize,
     AbortIfMultisigPayloadMismatch,
+    SimulationEnhancement,
```
Comment (banool) [inline aptos-move/aptos-release-builder/src/components/feature_flags.rs:124]: This is a bit vague, what if we enhance simulation again? Can we use a more descriptive name?

Also the AIP includes multiple changes, should they possibly be feature gated separately? I'm less opinionated on this one.

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 6]
```diff
@@ -2249,11 +2255,11 @@ impl AptosVM {
                     txn_data,
                     log_context,
                     traversal_context,
+                    self.is_simulation,
                 )?;
-                // Skip validation if this is part of tx simulation.
-                // This allows simulating multisig txs without having to first create the multisig
-                // tx.
-                if !self.is_simulation {
+                // Once "simulation_enhancement" is enabled, the simulation path also validates the
```
Comment (banool) [inline aptos-move/aptos-vm/src/aptos_vm.rs:2260]: Good comments, this and the code at 1146 make sense.

File: aptos-move/framework/aptos-framework/sources/transaction_validation.move [hunk 7]
```diff
@@ -123,24 +126,28 @@ module aptos_framework::transaction_validation {
                 error::invalid_argument(PROLOGUE_ESEQUENCE_NUMBER_TOO_NEW)
             );
 
-            assert!(
-                txn_authentication_key == bcs::to_bytes(&transaction_sender),
-                error::invalid_argument(PROLOGUE_EINVALID_ACCOUNT_AUTH_KEY),
-            );
+            if (!features::simulation_enhancement_enabled() || !is_simulation || !is_auth_key_empty(&txn_authentication_key)) {
```
Comment (banool) [inline aptos-move/framework/aptos-framework/sources/transaction_validation.move:129]: Nit: We do this same set of checks on line 99, consider a helper function.

File: aptos-move/framework/aptos-framework/sources/transaction_validation.move [hunk 8]
```diff
@@ -95,10 +96,12 @@ module aptos_framework::transaction_validation {
                 || txn_sequence_number > 0
         ) {
             assert!(account::exists_at(transaction_sender), error::invalid_argument(PROLOGUE_EACCOUNT_DOES_NOT_EXIST));
-            assert!(
-                txn_authentication_key == account::get_authentication_key(transaction_sender),
-                error::invalid_argument(PROLOGUE_EINVALID_ACCOUNT_AUTH_KEY),
-            );
+            if (!features::simulation_enhancement_enabled() || !is_simulation || !is_auth_key_empty(&txn_authentication_key)) {
```
Comment (banool) [inline aptos-move/framework/aptos-framework/sources/transaction_validation.move:99]: Maybe it's just me but I find !(a && b && c) simpler to reason about than (!a || !b || !c) hahah.

File: aptos-move/framework/aptos-framework/sources/transaction_validation.move [hunk 9]
```diff
@@ -263,10 +282,11 @@ module aptos_framework::transaction_validation {
         storage_fee_refunded: u64,
         txn_gas_price: u64,
         txn_max_gas_units: u64,
-        gas_units_remaining: u64
+        gas_units_remaining: u64,
+        is_simulation: bool,
```
Comment (banool) [inline aptos-move/framework/aptos-framework/sources/transaction_validation.move:286]: Not in this PR but it might be good to start using a struct for this, it's a lot of args.

File: aptos-move/framework/move-stdlib/sources/configs/features.move [hunk 10]
```diff
@@ -582,6 +582,18 @@ module std::features {
         is_enabled(ABORT_IF_MULTISIG_PAYLOAD_MISMATCH)
     }
 
+    /// Whether the multisig v2 fix is enabled. Once enabled, the multisig transaction execution will explicitly
+    /// abort if the provided payload does not match the payload stored on-chain.
```
Comment (banool) [inline aptos-move/framework/move-stdlib/sources/configs/features.move:586]: This also affects other stuff such as the authentication key and gas payer right?

File: types/src/transaction/authenticator.rs [hunk 11]
```diff
@@ -358,6 +358,9 @@ impl TransactionAuthenticator {
                 AccountAuthenticator::MultiKey { authenticator } => {
                     single_key_authenticators.extend(authenticator.to_single_key_authenticators()?);
                 },
+                AccountAuthenticator::NoAccountAuthenticator => {
+                    //  TODO: check this.
```
Comment (banool) [inline types/src/transaction/authenticator.rs:362]: No authenticator so there's nothing to do here?

**Review by banool — DISMISSED** [review]
We should have more tests for this path, in particular that the new variant can only be used in simulation. Move tests in the prologue module or something as well as API tests / smoke tests.

These prologue checks are getting quite complicated, might be good to get more eyes on this PR, esp from those familiar with the prologue.

File: aptos-move/framework/aptos-framework/sources/transaction_validation.move (hunk 9 above)
Comment (junkil-park) [reply aptos-move/framework/aptos-framework/sources/transaction_validation.move:286]: It's a good point!

File: aptos-move/framework/move-stdlib/sources/configs/features.move (hunk 10 above)
Comment (junkil-park) [reply aptos-move/framework/move-stdlib/sources/configs/features.move:586]: Updated!

File: aptos-move/aptos-vm/src/aptos_vm.rs (hunk 6 above)
Comment (junkil-park) [reply aptos-move/aptos-vm/src/aptos_vm.rs:2260]: Thanks!

File: api/src/tests/multisig_transactions_test.rs (hunk 3 above)
Comment (junkil-park) [reply api/src/tests/multisig_transactions_test.rs:414]: Although we have some 2-of-3 cases, I added some more (both positive and negative cases).

File: types/src/transaction/authenticator.rs (hunk 11 above)
Comment (junkil-park) [reply types/src/transaction/authenticator.rs]: Yes, no authentication is added in this case. I updated the comment accordingly. (cc: @heliuchuan )

File: api/src/tests/simulation_test.rs (hunk 4 above)
Comment (junkil-park) [reply api/src/tests/simulation_test.rs:223]: Added the tests here:
* https://github.com/aptos-labs/aptos-core/pull/13714/files#diff-3f5e53d39326e69c98cf9e1210c60a8a8e54e2668ef8e128c4d295fbc82964bdR389
* https://github.com/aptos-labs/aptos-core/pull/13714/files#diff-3f5e53d39326e69c98cf9e1210c60a8a8e54e2668ef8e128c4d295fbc82964bdR389

**Comment by junkil-park** [comment]
> Can we double check that you can't submit a NoSignature transaction and have it pass? Do we have new tests that cover this new path?

> We should have more tests for this path, in particular that the new variant can only be used in simulation. Move tests in the prologue module or something as well as API tests / smoke tests.

I added such tests here:
* https://github.com/aptos-labs/aptos-core/pull/13714/files#diff-3f5e53d39326e69c98cf9e1210c60a8a8e54e2668ef8e128c4d295fbc82964bdR389
* https://github.com/aptos-labs/aptos-core/pull/13714/files#diff-3f5e53d39326e69c98cf9e1210c60a8a8e54e2668ef8e128c4d295fbc82964bdR389

File: aptos-move/aptos-vm/src/transaction_metadata.rs [hunk 12]
```diff
@@ -37,20 +37,29 @@ impl TransactionMetadata {
     pub fn new(txn: &SignedTransaction) -> Self {
         Self {
             sender: txn.sender(),
-            authentication_key: txn.authenticator().sender().authentication_key().to_vec(),
+            authentication_key: txn
+                .authenticator()
+                .sender()
+                .authentication_key()
+                .map_or_else(Vec::new, |auth_key| auth_key.to_vec()),
```
Comment (movekevin) [inline aptos-move/aptos-vm/src/transaction_metadata.rs:44]: I assume this is to catch invalid formats of auth key. Does this swallow more errors that we don't want to swallow?

**Review by movekevin — COMMENTED** [review]
Can you:
1. Can you figure out a way to keep the same use case for multisig simulation where evs can simulate to check that a multisig tx can be executed BEFORE it's created? In this PR, since the simulation and execution paths are merged, this use case no longer works as the multisig tx has to be created BEFORE simulation can be done. One idea as discussed is to check if this use case is satisfied by simulating sending the multisig tx from the multisig account itself without an authenticator specified.
2. Add documentation and some basic tutorials for simulation now that things are much clearer and devs have more options when sending simulation requests.
3. Update SDK v2 in a separate PR (because separate repo).
4. Test more e2e by running a TS example with new simulation features against a local test node
5. Check that the gas estimation issue (under gassed) with executing multisig transactions is resolved with this change.

**Comment by junkil-park** [comment]
> Can you:
> 
> 1. Can you figure out a way to keep the same use case for multisig simulation where evs can simulate to check that a multisig tx can be executed BEFORE it's created? In this PR, since the simulation and execution paths are merged, this use case no longer works as the multisig tx has to be created BEFORE simulation can be done. One idea as discussed is to check if this use case is satisfied by simulating sending the multisig tx from the multisig account itself without an authenticator specified.

Yes, here is an example: https://github.com/aptos-labs/aptos-ts-sdk/pull/467/files#diff-19037adf2ba690a7d9cdfaea7398deae3906ec566ec7d49bbdf72c2453b82e2cR136 It's in a draft mode, and under discussion with Maayan.

> 2. Add documentation and some basic tutorials for simulation now that things are much clearer and devs have more options when sending simulation requests.

I added a description of the simulation behavior change in the AIP: https://github.com/aptos-foundation/AIPs/pull/468/files#diff-73fb9dc634df9f3142063c3d4b512113441bea69242015bc0988e58de38ba154R44. I will write a tutorial when the example above is finalized in the TS SDK side.

> 3. Update SDK v2 in a separate PR (because separate repo).
I have put up a PR here: https://github.com/aptos-labs/aptos-ts-sdk/pull/467. I am currently iterating it together with Maayan.

> 4. Test more e2e by running a TS example with new simulation features against a local test node
I am adding tests in the TS SDK update PR above:
* https://github.com/aptos-labs/aptos-ts-sdk/pull/467/files#diff-f97493b76e508e849a876ecb35492119c3ebf540140bda3c1a7b3313a32be2de
* https://github.com/aptos-labs/aptos-ts-sdk/pull/467/files#diff-f1a843f6e05220a54a794964cbcb789ee9cfd6671ef77518e3a4a75d891f30d7

> 5. Check that the gas estimation issue (under gassed) with executing multisig transactions is resolved with this change.
I checked it on the localnet and didn't encounter any gas issues after this change.

File: aptos-move/framework/aptos-framework/sources/transaction_validation.move (hunk 7 above)
Comment (junkil-park) [reply aptos-move/framework/aptos-framework/sources/transaction_validation.move]: Thank you! I refactored it with a helper function.

File: aptos-move/framework/aptos-framework/sources/transaction_validation.move (hunk 8 above)
Comment (junkil-park) [reply aptos-move/framework/aptos-framework/sources/transaction_validation.move]: Thank you! I refactored it with a helper function.

File: aptos-move/aptos-release-builder/src/components/feature_flags.rs (hunk 5 above)
Comment (junkil-park) [reply aptos-move/aptos-release-builder/src/components/feature_flags.rs:125]: Yes, this has multiple changes. We could rename it to `SimulationFlexibilityConsistencyEnhancement` to make it a bit more descriptive.

File: types/src/transaction/authenticator.rs [hunk 13]
```diff
@@ -358,6 +358,9 @@ impl TransactionAuthenticator {
                 AccountAuthenticator::MultiKey { authenticator } => {
                     single_key_authenticators.extend(authenticator.to_single_key_authenticators()?);
                 },
+                AccountAuthenticator::NoAccountAuthenticator => {
+                    //  This case adds no single key authenticators to the vector.
+                },
```
Comment (gregnazario) [inline types/src/transaction/authenticator.rs:363]: Does this get rejected by normal requests?

File: types/src/transaction/authenticator.rs [hunk 14]
```diff
@@ -578,12 +587,20 @@ impl AccountAuthenticator {
             Self::MultiEd25519 { signature, .. } => signature.to_bytes().to_vec(),
             Self::SingleKey { authenticator } => authenticator.signature_bytes(),
             Self::MultiKey { authenticator } => authenticator.signature_bytes(),
+            Self::NoAccountAuthenticator => vec![], // TODO: check this.
```
Comment (gregnazario) [inline types/src/transaction/authenticator.rs:590]: todo?

File: aptos-move/framework/aptos-framework/sources/transaction_validation.move [hunk 15]
```diff
@@ -290,43 +315,53 @@ module aptos_framework::transaction_validation {
...
         let addr = signer::address_of(&account);
         account::increment_sequence_number(addr);
     }
+
+    inline fun skip_auth_key_check(is_simulation: bool, auth_key: &vector<u8>): bool {
+        is_simulation && vector::is_empty(auth_key)
+    }
+
+    inline fun skip_gas_payment(is_simulation: bool, gas_payer: address): bool {
+        is_simulation && gas_payer == @0x0
```
Comment (gregnazario) [inline aptos-move/framework/aptos-framework/sources/transaction_validation.move:365]: So, I might have missed it, but there's no way for someone to override the is_simulation at all in a real VM right?

**Review by movekevin — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/transaction_validation.move (hunk 15 above)
Comment (junkil-park) [reply aptos-move/framework/aptos-framework/sources/transaction_validation.move:365]: `is_simulation` can be true only inside of `AptosSimulationVM` (`pub struct AptosSimulationVM(AptosVM);`).
https://github.com/aptos-labs/aptos-core/blob/07e9f7866b88857f0ca1d7ea6f49529575bf9617/aptos-move/aptos-vm/src/aptos_vm.rs#L2721C12-L2721C32

File: types/src/transaction/authenticator.rs (hunk 14 above)
Comment (junkil-park) [reply types/src/transaction/authenticator.rs:590]: `NoAccountAuthenticator` contains neither a public key nor a signature. So, an empty vector is returned. I removed the TODOs.

File: types/src/transaction/authenticator.rs (hunk 13 above)
Comment (junkil-park) [reply types/src/transaction/authenticator.rs:363]: The use of `NoAccountAuthenticator` should result in a failure during the signature verification of the transaction execution:
```
    pub fn verify<T: Serialize + CryptoHash>(&self, message: &T) -> Result<()> {
        match self {
            ...
            Self::NoAccountAuthenticator => bail!("No signature to verify."),
        }
    }
```
(https://github.com/aptos-labs/aptos-core/pull/13714/files#diff-770742fc2aa984d2ad09f991d7dd190d3e0888189a87e3d3d25768d6d1939dbeR568)

**Review by gregnazario — APPROVED** [review]
